PolyEngine/API/Src/pe/api/deps/DependencyManager.cpp

```diff
@@ -0,0 +1,124 @@
+#include <pe/api/IEngine.hpp>
+
+#include <pe/api/deps/DependencyManager.hpp>
+#include <pe/core/utils/Logger.hpp>
+
+using ::pe::core::utils::gConsole;
+
+namespace pe::api::deps {
+
+DependencyManager& DependencyManager::get()
+{
+	static DependencyManager instance;
+	return instance;
+}
+
+DependencyManager::DependencyManager()
+{
+	gConsole.LogInfo("[DependencyManager] Dependency manager initialized.");
+}
+
+DependencyManager::~DependencyManager()
+{
+	gConsole.LogInfo("[DependencyManager] Dependency manager deinitialized.");
+}
+
+void DependencyManager::registerDependency(IDependency* dependency)
+{
+	ASSERTE(dependency, "Cannot register nullptr");
+	auto type = dependency->getType();
+	auto it = m_dependencyMap.find(type);
+	ASSERTE(it == m_dependencyMap.end(), "Registering dependency of given type twice in a row.");
+	m_dependencyMap.emplace(dependency->getType(), dependency);
+
+	gConsole.LogInfo("[DependencyManager] Registered new dependency [{}].", dependency->getName());
+
+	injectAll(dependency);
+}
+
+void DependencyManager::unregisterDependency(IDependency* dependency)
+{
+	ASSERTE(dependency, "Cannot unregister nullptr");
+	auto type = dependency->getType();
+	auto it = m_dependencyMap.find(type);
+	ASSERTE(it != m_dependencyMap.end(), "Cannot unregister non-registered dependency!");
+	ASSERTE(it->second == dependency, "Cannot unregister different dependency!");
```

Comment:
Can this even happen?

Reply:
It shouldn't if the code is written correctly. Assertions should be used for things that should not ever happen unless there is a bug in the code.